File: app/src/main/java/com/example/leo/test/MediaListViewModel.java
package com.example.leo.test;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;
import android.content.Context;
import android.database.Cursor;
import android.provider.MediaStore;

import java.util.ArrayList;
import java.util.List;

public class MediaListViewModel extends ViewModel {
    private MutableLiveData<List<Media>> mMedias = new MutableLiveData<>();
    private int mCurrentSelectedMediaIndex;

    public LiveData<List<Media>> getAllMedias(Context context) {
        fetchMedias(context);
        return mMedias;
    }

    public void select(String mediaUrl) {
        for (int i = 0; i < mMedias.getValue().size(); i++) {
            if (mediaUrl.equals(mMedias.getValue().get(i).url)) {
                mCurrentSelectedMediaIndex = i;
                return;
            }
        }
    }

    public String moveToNextMedia() {
        mCurrentSelectedMediaIndex = (mCurrentSelectedMediaIndex + 1) % mMedias.getValue().size();
        return getCurrentMedia();
    }

    public String moveToPreviousMedia() {
        mCurrentSelectedMediaIndex = mCurrentSelectedMediaIndex == 0 ? mMedias.getValue().size() - 1
                : mCurrentSelectedMediaIndex - 1;

        return getCurrentMedia();
    }

    public String getCurrentMedia() {
        return mMedias.getValue().get(mCurrentSelectedMediaIndex).url;
    }

    private void fetchMedias(Context context) {
        Cursor cursor = context.getContentResolver().query(
                MediaStore.Audio.Media.EXTERNAL_CONTENT_URI,
                null,
                MediaStore.Audio.Media.IS_MUSIC + "<>0", null,
                MediaStore.Audio.Media.DEFAULT_SORT_ORDER
        );

        List<Media> allMedias = new ArrayList<>(cursor.getCount());

        try {
            while (cursor.moveToNext()) {
                Media media = new Media();
                media.id = cursor.getInt(cursor.getColumnIndex(MediaStore.Audio.Media._ID));
                media.title = cursor.getString(cursor.getColumnIndex(MediaStore.Audio.Media.TITLE));
                media.artist = cursor.getString(cursor.getColumnIndex(MediaStore.Audio.Media.ARTIST));
                media.url = cursor.getString(cursor.getColumnIndex(MediaStore.Audio.Media.DATA));

                allMedias.add(media);
            }
        } finally {
            cursor.close();
        }

        mMedias.setValue(allMedias);
    }
}

File: app/src/main/kotlin/com/example/leo/test/MediaListViewModel.kt
package com.example.leo.test

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import android.content.Context
import android.provider.MediaStore
import java.util.*

class MediaListViewModel : ViewModel() {
    private val medias = MutableLiveData<List<Media>>()
    private var currentSelectedMediaIndex = 0

    fun getAllMedias(context: Context): LiveData<List<Media>> {
        fetchMedias(context)
        return medias
    }

    fun select(mediaUrl: String) = medias.value?.let {
        for (i in 0 until it.size) {
            if (mediaUrl == it[i].url) {
                currentSelectedMediaIndex = i
                break
            }
        }
    }

    fun nextMedia() = medias.value?.let {
        val nextIndex = (currentSelectedMediaIndex + 1) % it.size
        it[nextIndex].url
    }

    fun previousMedia() = medias.value?.let {
        val previousIndex = if (currentSelectedMediaIndex == 0)
            it.size - 1
        else
            currentSelectedMediaIndex - 1

        it[previousIndex].url
    }

    fun currentMedia() = medias.value?.get(currentSelectedMediaIndex)

    private fun fetchMedias(context: Context) {
        val cursor = context.contentResolver.query(
                MediaStore.Audio.Media.EXTERNAL_CONTENT_URI, null,
                MediaStore.Audio.Media.IS_MUSIC + "<>0", null,
                MediaStore.Audio.Media.DEFAULT_SORT_ORDER
        )

        cursor?.use {
            val allMedias = ArrayList<Media>(cursor.count)

            while (it.moveToNext()) {
                val media = Media().apply {
                    id = it.getInt(it.getColumnIndex(MediaStore.Audio.Media._ID))
                    title = it.getString(it.getColumnIndex(MediaStore.Audio.Media.TITLE))
                    artist = it.getString(it.getColumnIndex(MediaStore.Audio.Media.ARTIST))
                    url = it.getString(it.getColumnIndex(MediaStore.Audio.Media.DATA))
                }

                allMedias.add(media)
            }
            medias.value = allMedias
        }
    }
}
